### PR TITLE
PDF and Email Sending of Receipts

### LINKED ISSUE #25: PDF and Email Sending of Receipts
- view
 - urls
 - weasyprint
 - sendgrid

### PR DESCRIPTION
**Pull request by aleisley**
- This resolves #25 (PDF and Email Sending of Receipts)

### CODE REVIEW

File: payroll_system/receipts/views.py [hunk 1]
```diff
@@ -36,4 +47,42 @@ def get_redirect_url(self, *args, **kwargs):
         receipt.save()
 
         # Redirect to detail page
-        return reverse('receipts:detail', kwargs={'pk': receipt.pk})
\ No newline at end of file
+        return reverse('receipts:detail', kwargs={'pk': receipt.pk})
+
+
+class GeneratePdfView(LoginRequiredMixin, View):
```
Comment (antoniocarlosortiz) [inline payroll_system/receipts/views.py]: make Pdf all caps

File: payroll_system/receipts/views.py [hunk 2]
```diff
@@ -36,4 +47,42 @@ def get_redirect_url(self, *args, **kwargs):
...
+    '''
+    Generates the template pdf for the payslip making use of the
+    generate_employee_payroll_receipt method of the Payroll object
+    '''
+
+    def get(self, request, *args, **kwargs):
+        # Get receipt object generated from Payroll object method
+        receipt_id = self.kwargs['pk']
+        receipt = Receipt.objects.get(pk=receipt_id)
```
Comment (antoniocarlosortiz) [inline payroll_system/receipts/views.py]: the receipt can be hidden or not. Make sure this does not work if its hidden

File: payroll_system/receipts/tasks.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+from weasyprint import CSS
+from weasyprint import HTML
+from weasyprint.fonts import FontConfiguration
+
+from .models import Receipt
+
+
+@shared_task
+def send_pdf_to_mail(receipt_id):
```
Comment (antoniocarlosortiz) [inline payroll_system/receipts/tasks.py]: this should be a model method of the Receipt model

File: payroll_system/receipts/templates/receipts/payslip.html [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+    <meta http-equiv="X-UA-Compatible" content="ie=edge">
+    <meta name="description" content="Payslip">
+    <meta name="authors" content="Anthon/Ivan">
+    
+    <!-- <link href="#" media="print" rel="stylesheet"> -->
+    <title>Payslip</title>
+
+    <style>
+        html{
```
Comment (antoniocarlosortiz) [inline payroll_system/receipts/templates/receipts/payslip.html]: add a space before the {

File: payroll_system/receipts/views.py (hunk 1 above)
Comment (aleisley) [reply payroll_system/receipts/views.py]: Class name now GeneratePDFView

File: payroll_system/receipts/views.py (hunk 2 above)
Comment (aleisley) [reply payroll_system/receipts/views.py]: Added is_hidden and deleted_dt fields in Receipt model. Here, receipt is now retrieved with get_object_or_404, with is_hidden required to be False.

File: payroll_system/receipts/tasks.py (hunk 3 above)
Comment (aleisley) [reply payroll_system/receipts/tasks.py]: Function above is transferred to the Receipt model as a method. In tasks.py, the function is renamed as async_receipt_mail and now calls the model method in Receipt.

File: payroll_system/receipts/templates/receipts/payslip.html (hunk 4 above)
Comment (aleisley) [reply payroll_system/receipts/templates/receipts/payslip.html]: Added spaces before { in CSS styles.

File: payroll_system/receipts/models.py [hunk 5]
```diff
@@ -48,9 +56,50 @@ class Receipt(models.Model):
...
+
+        font_config = FontConfiguration()
+        pdf = HTML(string=html).write_pdf(
+            font_config=font_config
+        )
+
+        message = EmailMultiAlternatives(**email_details)
+
+        message.attach(f'Payslip for {self}', pdf, 'application/pdf')
+        message.send()
```
[the remark this answers is not in the document]
Comment (aleisley) [reply payroll_system/receipts/models.py]: Sir @antoniocarlosortiz, the return 0 is down below... at line number 107.

File: payroll_system/receipts/tasks.py [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+
+
+@shared_task
+def async_receipt_mail(receipt_id):
+    '''
+    Utilizes the model method in the receipt to send mail
+    asynchronously.
+    '''
+    receipt = get_object_or_404(Receipt, pk=receipt_id, is_hidden=False)
+    receipt.send_pdf_to_mail()
```
[the remark this answers is not in the document]
Comment (aleisley) [reply payroll_system/receipts/tasks.py]: Oh ok... we thought to only add it in the model method. Added return 0 below receipt.send_pdf_to_mail

**Pull request merged**